### PR TITLE
Fix allreduce dependency order.

### PR DESCRIPTION
**Pull request by gongweibao**

### CODE REVIEW

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+  return nullptr;
+}
+
+std::unique_ptr<ir::Graph> AllReduceDepsPass::ApplyImpl(
+    std::unique_ptr<ir::Graph> graph) const {
+  auto graph_ops = ir::FilterByNodeWrapper<OpHandleBase>(*graph);
+
+  // get vars level
+  int level = 0;
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc]: better name `order` to understand

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+  std::unordered_map<std::string, int> vars;
+  // TODO(gongwb): use graph topology sort to find the order of operators.
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc:53]: Note that must assert topology sort is stable

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        vars[v] = level;
+      }
+    }
+    level++;
+  }
+
+  std::vector<OpHandleBase*> dist_ops;
+  // get allreduce ops.
+  for (auto& op : graph_ops) {
+    if (op->Name() == "all_reduce" || op->Name() == "reduce") {
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/all_reduce_deps_pass.cc:70]: broadcast?

File: paddle/fluid/framework/details/build_strategy.cc [hunk 4]
```diff
@@ -123,7 +132,14 @@ std::unique_ptr<ir::Graph> BuildStrategy::Apply(
       pass->Erase("nccl_ctxs");
       pass->SetNotOwned<platform::NCCLContextMap>("nccl_ctxs", nctx);
 #endif
-    } else if (pass->Type() == "sequential_execution_pass") {
+    } else if (pass->Type() == "all_reduce_deps_pass" ||
```
Comment (typhoonzero) [inline paddle/fluid/framework/details/build_strategy.cc]: should separate these two

File: python/paddle/fluid/parallel_executor.py [hunk 5]
```diff
@@ -124,16 +124,12 @@ def __init__(self,
...
-        #   env to make sure all gpus run ops in same order.
-        if num_trainers > 1:
-            assert (use_cuda)
-            # FIXME(gongwb): avoid this set.
-            exec_strategy.num_threads = 1
-
         if build_strategy is None:
             build_strategy = BuildStrategy()
 
+        build_strategy.num_trainers = num_trainers
```
Comment (typhoonzero) [inline python/paddle/fluid/parallel_executor.py:130]: if adding these deps do not harm the performance, can run this pass by default, and no need to add these attributes.

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc (hunk 1 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_deps_pass.cc]: Done.

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_deps_pass.cc:53]: Thanks. Add to Comment.

File: python/paddle/fluid/parallel_executor.py (hunk 5 above)
Comment (gongweibao) [reply python/paddle/fluid/parallel_executor.py:130]: It harms the performance a little now.

File: paddle/fluid/framework/details/all_reduce_deps_pass.cc (hunk 3 above)
Comment (gongweibao) [reply paddle/fluid/framework/details/all_reduce_deps_pass.cc:70]: Oh, it's easy to add circle if I add broadcast.
I'll fix it later.https://github.com/PaddlePaddle/Paddle/issues/14597

**Comment by Yancey0623** [comment]
Maybe we don't need a new pass, my immature idea:
1. wapper all ther nccl op handle to append nccl operator
1. add the dep var for the new nccl op in the wapper function, and the dep var is the member of `multi_devices_pass`.


advantage: The code is more clearly, don't need to analyized the graph.

**Comment by gongweibao** [comment]
Thanks. I'll test it later.

File: python/paddle/fluid/parallel_executor.py (hunk 5 above)
Comment (typhoonzero) [reply python/paddle/fluid/parallel_executor.py:130]: We need to make this API easy to understand later, for now, ps mode should not pass num_trainers.

**Review by typhoonzero — APPROVED** [review]
LGTM with one note.

**Comment by typhoonzero** [comment]
Need unitest, please add ut ASAP

**Pull request merged**